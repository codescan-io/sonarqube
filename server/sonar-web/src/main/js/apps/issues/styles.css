/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.issues-main-header .component-name {
  line-height: var(--controlHeight);
}

.issues-main-header-spinner {
  margin-right: 2px;
}

.concise-issues-list-header-inner {
  width: 260px;
  text-align: center;
}

.concise-issues-list-header .spinner {
  margin-top: 4px;
  margin-left: 1px;
  margin-right: 1px;
}

.concise-issues-list-header-button {
  border: none;
}

.concise-issues-list-header-button path {
  fill: var(--secondFontColor);
  transition: fill 0.3s ease;
}

.concise-issues-list-header-button:hover path {
  fill: var(--blue);
}

.concise-issue-component {
  margin-top: calc(var(--gridSize) * 2);
  margin-bottom: calc(var(--gridSize) / 2);
  padding-left: var(--gridSize);
  padding-right: var(--gridSize);
  direction: rtl;
}

.concise-issue-box {
  position: relative;
  z-index: var(--belowNormalZIndex);
  margin-bottom: calc(var(--gridSize) / 2);
  padding: calc(var(--gridSize) - 1px);
  border: 2px solid var(--barBorderColor);
  background-color: #fff;
  cursor: pointer;
  transition: background-color 0.3s ease, border-color 0.3s ease;
}

.not-all-issue-warning {
  padding: 16px 16px 0;
  width: 100%;
  box-sizing: border-box;
}

.not-all-issue-warning.open-issue-list {
  position: sticky;
  top: 0;
  z-index: 1000;
  background-color: var(--barBackgroundColor);
  display: inline-block;
}

.concise-issue-box .issue-message-highlight-CODE {
  background-color: var(--blacka06);
}

.concise-issue-box:hover {
  border: 2px dashed var(--blue);
}

.concise-issue-box:focus {
  outline: none;
}

.concise-issue-box.selected {
  z-index: var(--normalZIndex);
  border: 2px solid var(--blue);
  cursor: default;
}

.concise-issue-box-message,
.concise-issue-box-message:focus,
.concise-issue-box-message:hover,
.concise-issue-box-message:active {
  display: block;
  word-break: break-word;
  font-weight: bold !important;
  color: inherit !important;
  text-align: left;
  max-width: 100%;
}

.concise-issue-box-attributes {
  margin-top: var(--gridSize);
  line-height: 16px;
  font-size: var(--smallFontSize);
  display: flex;
  align-items: flex-start;
  flex-wrap: wrap;
  justify-content: flex-start;
}

.concise-issue-box-attributes .concise-issue-box-flow-indicator {
  margin-left: auto;
}

.concise-issue-box:not(.selected) .location-index {
  background-color: var(--secondFontColor);
}

.concise-issue-locations {
  display: inline-block;
  margin-bottom: -4px;
  margin-left: var(--gridSize);
}

.concise-issue-box-attributes > .location-index {
  margin-bottom: 4px;
  margin-right: 4px;
}

.concise-issue-box-attributes > .concise-issue-expand {
  background-color: transparent;
  border: 1px solid var(--conciseIssueRed);
  height: 16px;
  color: var(--conciseIssueRed);
  font-weight: bold;
  font-size: 16px;
  line-height: 8px;
  padding-bottom: 6px;
}

.concise-issue-box-attributes > .concise-issue-expand:hover {
  background-color: var(--conciseIssueRed);
  color: white;
}

.issues-page-actions {
  display: inline-block;
  min-width: 80px;
  text-align: right;
}

.issues .issue-list {
  /* no math, just a good guess */
  min-width: 640px;
  width: 800px;
}

.issues .issue a:focus,
.issues .issue button:focus {
  box-shadow: none;
}

@media (max-width: 1320px) {
  .issues .issue-list {
    width: calc(60vw - 40px);
  }
}

.issue-header .issue-message-highlight-CODE {
  background-color: var(--blacka06);
  border-radius: 5px;
}

.issue-location {
  display: inline-block;
  vertical-align: top;
  line-height: 18px;
  height: 18px;
  box-sizing: border-box;
  background-color: var(--issueBgColor);
  transition: background-color 0.3s ease;
}

.issues-workspace-list-component {
  padding: 10px 0 6px;
}

.issues-workspace-list-item + .issues-workspace-list-item {
  margin-top: 5px;
}

.issues-workspace-list-component + .issues-workspace-list-item {
  margin-top: 10px;
}

.issues-workspace-list-item:first-child .issues-workspace-list-component {
  padding-top: 0;
}

.issues-workspace-list-component + .issues-workspace-list-item {
  margin-top: 0;
}

.issues-predefined-periods {
  display: flex;
}

.issues-predefined-periods .search-navigator-facet {
  width: auto;
  margin-right: calc(var(--gridSize) / 2);
}

.bulk-change-radio-button {
  margin: 0 calc(-1 * var(--gridSize) / 2);
  padding: 0 calc(var(--gridSize) / 2);
}

.bulk-change-radio-button:hover {
  background-color: var(--barBackgroundColor);
}

.layout-page-main.open-issue {
  padding-top: 0;
}
