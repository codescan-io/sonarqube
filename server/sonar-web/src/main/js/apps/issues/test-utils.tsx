/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { waitFor } from '@testing-library/react';
import React from 'react';
import { byLabelText, byRole } from 'testing-library-selector';
import ComponentsServiceMock from '../../api/mocks/ComponentsServiceMock';
import IssuesServiceMock from '../../api/mocks/IssuesServiceMock';
import { mockComponent } from '../../helpers/mocks/component';
import { mockCurrentUser } from '../../helpers/testMocks';
import { renderApp, renderAppWithComponentContext } from '../../helpers/testReactTestingUtils';
import { Component } from '../../types/types';
import { CurrentUser } from '../../types/users';
import IssuesApp from './components/IssuesApp';
import { projectIssuesRoutes } from './routes';

jest.mock('../../api/issues');
jest.mock('../../api/rules');
jest.mock('../../api/components');
jest.mock('../../api/users');

export const issuesHandler = new IssuesServiceMock();
export const componentsHandler = new ComponentsServiceMock();

export const ui = {
  loading: byLabelText('loading'),
  issueItems: byRole('region'),

  issueItem1: byRole('region', { name: 'Issue with no location message' }),
  issueItem2: byRole('region', { name: 'FlowIssue' }),
  issueItem3: byRole('region', { name: 'Issue on file' }),
  issueItem4: byRole('region', { name: 'Fix this' }),
  issueItem5: byRole('region', { name: 'Fix that' }),
  issueItem6: byRole('region', { name: 'Second issue' }),
  issueItem7: byRole('region', { name: 'Issue with tags' }),
  issueItem8: byRole('region', { name: 'Issue on page 2' }),

  clearFitForDevelopmentFacet: byRole('button', {
    name: 'clear_x_filter.issues.facet.characteristics.DEVELOPMENT',
  }),
  clearCharacteristicFilter: byRole('checkbox', { name: 'issue.characteristic.CLEAR' }),
  typeFacet: byRole('button', { name: 'issues.facet.types' }),
  clearIssueTypeFacet: byRole('button', { name: 'clear_x_filter.issues.facet.types' }),
  codeSmellIssueTypeFilter: byRole('checkbox', { name: 'issue.type.CODE_SMELL' }),
  vulnerabilityIssueTypeFilter: byRole('checkbox', { name: 'issue.type.VULNERABILITY' }),
  clearSeverityFacet: byRole('button', { name: 'clear_x_filter.issues.facet.severities' }),
  majorSeverityFilter: byRole('checkbox', { name: 'severity.MAJOR' }),
  scopeFacet: byRole('button', { name: 'issues.facet.scopes' }),
  clearScopeFacet: byRole('button', { name: 'clear_x_filter.issues.facet.scopes' }),
  mainScopeFilter: byRole('checkbox', { name: 'issue.scope.MAIN' }),
  resolutionFacet: byRole('button', { name: 'issues.facet.resolutions' }),
  clearResolutionFacet: byRole('button', { name: 'clear_x_filter.issues.facet.resolutions' }),
  fixedResolutionFilter: byRole('checkbox', { name: 'issue.resolution.FIXED' }),
  statusFacet: byRole('button', { name: 'issues.facet.statuses' }),
  creationDateFacet: byRole('button', { name: 'issues.facet.createdAt' }),
  clearCreationDateFacet: byRole('button', { name: 'clear_x_filter.issues.facet.createdAt' }),
  clearStatusFacet: byRole('button', { name: 'clear_x_filter.issues.facet.statuses' }),
  openStatusFilter: byRole('checkbox', { name: 'issue.status.OPEN' }),
  confirmedStatusFilter: byRole('checkbox', { name: 'issue.status.CONFIRMED' }),
  languageFacet: byRole('button', { name: 'issues.facet.languages' }),
  ruleFacet: byRole('button', { name: 'issues.facet.rules' }),
  clearRuleFacet: byRole('button', { name: 'clear_x_filter.issues.facet.rules' }),
  tagFacet: byRole('button', { name: 'issues.facet.tags' }),
  clearTagFacet: byRole('button', { name: 'clear_x_filter.issues.facet.tags' }),
  projectFacet: byRole('button', { name: 'issues.facet.projects' }),
  clearProjectFacet: byRole('button', { name: 'clear_x_filter.issues.facet.projects' }),
  assigneeFacet: byRole('button', { name: 'issues.facet.assignees' }),
  clearAssigneeFacet: byRole('button', { name: 'clear_x_filter.issues.facet.assignees' }),
  authorFacet: byRole('button', { name: 'issues.facet.authors' }),
  clearAuthorFacet: byRole('button', { name: 'clear_x_filter.issues.facet.authors' }),

  dateInputMonthSelect: byRole('combobox', { name: 'Month:' }),
  dateInputYearSelect: byRole('combobox', { name: 'Year:' }),

  clearAllFilters: byRole('button', { name: 'clear_all_filters' }),

  ruleFacetList: byRole('list', { name: 'rules' }),
  languageFacetList: byRole('list', { name: 'languages' }),
  ruleFacetSearch: byRole('searchbox', { name: 'search.search_for_rules' }),
};

export async function waitOnDataLoaded() {
  await waitFor(() => {
    expect(ui.loading.query()).not.toBeInTheDocument();
  });
}

export function renderIssueApp(currentUser?: CurrentUser) {
  renderApp('project/issues', <IssuesApp />, { currentUser: mockCurrentUser(currentUser) });
}

export function renderProjectIssuesApp(navigateTo?: string, overrides?: Partial<Component>) {
  renderAppWithComponentContext(
    'project/issues',
    projectIssuesRoutes,
    { navigateTo },
    { component: mockComponent(overrides) }
  );
}
