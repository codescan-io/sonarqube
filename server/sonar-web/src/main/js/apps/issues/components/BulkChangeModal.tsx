/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import {
  ButtonPrimary,
  Checkbox,
  DeferredSpinner,
  FlagMessage,
  FormField,
  HelperHintIcon,
  Highlight,
  InputSelect,
  LabelValueSelectOption,
  LightLabel,
  Modal,
  RadioButton,
} from 'design-system';
import { pickBy, sortBy } from 'lodash';
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { SingleValue } from 'react-select';
import { bulkChangeIssues, searchIssueTags } from '../../../api/issues';
import FormattingTips from '../../../components/common/FormattingTips';
import IssueSeverityIcon from '../../../components/icon-mappers/IssueSeverityIcon';
import IssueTypeIcon from '../../../components/icon-mappers/IssueTypeIcon';
import { SEVERITIES } from '../../../helpers/constants';
import { throwGlobalError } from '../../../helpers/error';
import { translate, translateWithParameters } from '../../../helpers/l10n';
import { IssueSeverity } from '../../../types/issues';
import { Dict, Issue, IssueType, Paging } from '../../../types/types';
import AssigneeSelect from './AssigneeSelect';
import TagsSelect from './TagsSelect';
import { withOrganizationContext } from "../../organizations/OrganizationContext";

const DEBOUNCE_DELAY = 250;

interface Props {
  fetchIssues: (x: {}) => Promise<{ issues: Issue[]; paging: Paging }>;
  onClose: () => void;
  onDone: () => void;
  organization: Organization;
}

interface FormFields {
  addTags?: Array<string>;
  assignee?: SingleValue<LabelValueSelectOption<string>>;
  comment?: string;
  notifications?: boolean;
  organization: Organization;
  removeTags?: Array<string>;
  severity?: string;
  transition?: string;
  type?: string;
}

interface State extends FormFields {
  initialTags: Array<string>;
  issues: Issue[];
  // used for initial loading of issues
  loading: boolean;
  paging?: Paging;
  // used when submitting a form
  submitting: boolean;
}

enum InputField {
  addTags = 'addTags',
  assignee = 'assignee',
  removeTags = 'removeTags',
  severity = 'severity',
  type = 'type',
}

export const MAX_PAGE_SIZE = 500;

export class BulkChangeModal extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);
    this.state = { initialTags: [], issues: [], loading: true, submitting: false };
  }

  componentDidMount() {
    this.mounted = true;

    Promise.all([this.loadIssues(), searchIssueTags({ organization: this.props.organization.kee })]).then(
      ([{ issues, paging }, tags]) => {
        if (this.mounted) {
          if (issues.length > MAX_PAGE_SIZE) {
            issues = issues.slice(0, MAX_PAGE_SIZE);
          }

          this.setState({
            initialTags: tags,
            issues,
            loading: false,
            paging,
          });
        }
      },
      () => {}
    );
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  loadIssues = () => {
    return this.props.fetchIssues({ additionalFields: 'actions,transitions', ps: MAX_PAGE_SIZE });
  };

  handleAssigneeSelect = (assignee: SingleValue<LabelValueSelectOption<string>>) => {
    this.setState({ assignee });
  };

  handleTagsSearch = (query: string): Promise<string[]> => {
    return searchIssueTags({ q: query, organization: this.props.organization.kee })
      .then((tags) => tags)
      .catch(() => []);
  };

  handleTagsSelect =
    (field: InputField.addTags | InputField.removeTags) => (options: Array<string>) => {
      this.setState<keyof FormFields>({ [field]: options });
    };

  handleFieldCheck = (field: keyof FormFields) => (checked: boolean) => {
    if (!checked) {
      this.setState<keyof FormFields>({ [field]: undefined });
    } else if (field === 'notifications') {
      this.setState<keyof FormFields>({ [field]: true });
    }
  };

  handleRadioTransitionChange = (transition: string) => {
    this.setState({ transition });
  };

  handleCommentChange = (event: React.SyntheticEvent<HTMLTextAreaElement>) => {
    this.setState({ comment: event.currentTarget.value });
  };

  handleSelectFieldChange =
    (field: 'severity' | 'type') => (data: LabelValueSelectOption<string> | null) => {
      if (data) {
        this.setState<keyof FormFields>({ [field]: data.value });
      } else {
        this.setState<keyof FormFields>({ [field]: undefined });
      }
    };

  handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const query = pickBy(
      {
        add_tags: this.state.addTags?.join(),
        assign: this.state.assignee ? this.state.assignee.value : null,
        comment: this.state.comment,
        do_transition: this.state.transition,
        remove_tags: this.state.removeTags?.join(),
        sendNotifications: this.state.notifications,
        set_severity: this.state.severity,
        set_type: this.state.type,
      },
      (x) => x !== undefined
    );

    const issueKeys = this.state.issues.map((issue) => issue.key);

    this.setState({ submitting: true });
    bulkChangeIssues(issueKeys, query).then(
      () => {
        this.setState({ submitting: false });
        this.props.onDone();
      },
      (error) => {
        this.setState({ submitting: false });
        throwGlobalError(error);
      }
    );
  };

  getAvailableTransitions(issues: Issue[]) {
    const transitions: Dict<number> = {};
    issues.forEach((issue) => {
      if (issue.transitions) {
        issue.transitions.forEach((t) => {
          if (transitions[t] !== undefined) {
            transitions[t]++;
          } else {
            transitions[t] = 1;
          }
        });
      }
    });
    return sortBy(Object.keys(transitions)).map((transition) => ({
      transition,
      count: transitions[transition],
    }));
  }

  canSubmit = () => {
    const { addTags, assignee, removeTags, severity, transition, type } = this.state;

    return Boolean(
      (addTags && addTags.length > 0) ||
        (removeTags && removeTags.length > 0) ||
        assignee ||
        severity ||
        transition ||
        type
    );
  };

  renderField = (
    field: InputField,
    label: string,
    affected: number | undefined,
    input: React.ReactNode
  ) => (
    <FormField htmlFor={`issues-bulk-change-${field}`} label={translate(label)}>
      <div className="sw-flex sw-items-center sw-justify-between">
        {input}
        {affected !== undefined && (
          <LightLabel>
            ({translateWithParameters('issue_bulk_change.x_issues', affected)})
          </LightLabel>
        )}
      </div>
    </FormField>
  );

  renderAssigneeField = () => {
    const { assignee, issues } = this.state;
    const affected = this.state.issues.filter(hasAction('assign')).length;
    const field = InputField.assignee;

    if (affected === 0) {
      return null;
    }

    const input = (
      <AssigneeSelect
        assignee={assignee}
        className="sw-max-w-abs-300"
        inputId={`issues-bulk-change-${field}`}
        issues={issues}
        organization={this.props.organization.kee}
        onAssigneeSelect={this.handleAssigneeSelect}
      />
    );

    return this.renderField(field, 'issue.assign.formlink', affected, input);
  };

  renderTypeField = () => {
    const affected = this.state.issues.filter(hasAction('set_type')).length;
    const field = InputField.type;

    if (affected === 0) {
      return null;
    }

    const types: IssueType[] = ['BUG', 'VULNERABILITY', 'CODE_SMELL'];
    const options: LabelValueSelectOption<IssueType>[] = types.map((type) => ({
      label: translate('issue.type', type),
      value: type,
      Icon: <IssueTypeIcon height={16} type={type} />,
    }));

    const input = (
      <InputSelect
        className="sw-w-abs-300"
        inputId={`issues-bulk-change-${field}`}
        isClearable
        isSearchable={false}
        onChange={this.handleSelectFieldChange('type')}
        options={options}
        size="full"
      />
    );

    return this.renderField(field, 'issue.set_type', affected, input);
  };

  renderSeverityField = () => {
    const affected = this.state.issues.filter(hasAction('set_severity')).length;
    const field = InputField.severity;

    if (affected === 0) {
      return null;
    }

    const options: LabelValueSelectOption<IssueSeverity>[] = SEVERITIES.map((severity) => ({
      label: translate('severity', severity),
      value: severity,
      Icon: <IssueSeverityIcon height={16} severity={severity} />,
    }));

    const input = (
      <InputSelect
        className="sw-w-abs-300"
        inputId={`issues-bulk-change-${field}`}
        isClearable
        isSearchable={false}
        onChange={this.handleSelectFieldChange('severity')}
        options={options}
        size="full"
      />
    );

    return this.renderField(field, 'issue.set_severity', affected, input);
  };

  renderTagsField = (
    field: InputField.addTags | InputField.removeTags,
    label: string,
    allowCreate: boolean
  ) => {
    const { initialTags } = this.state;
    const tags = this.state[field] ?? [];
    const affected = this.state.issues.filter(hasAction('set_tags')).length;

    if (initialTags === undefined || affected === 0) {
      return null;
    }

    const input = (
      <TagsSelect
        allowCreation={allowCreate}
        inputId={`issues-bulk-change-${field}`}
        onChange={this.handleTagsSelect(field)}
        selectedTags={tags}
        onSearch={this.handleTagsSearch}
      />
    );

    return this.renderField(field, label, affected, input);
  };

  renderTransitionsField = () => {
    const transitions = this.getAvailableTransitions(this.state.issues);

    if (transitions.length === 0) {
      return null;
    }

    return (
      <div className="sw-mb-6">
        <fieldset>
          <Highlight as="legend" className="sw-mb-2">
            {translate('issue.transition')}
          </Highlight>
          {transitions.map((transition) => (
            <div
              className="sw-mb-1 sw-flex sw-items-center sw-justify-between"
              key={transition.transition}
            >
              <RadioButton
                checked={this.state.transition === transition.transition}
                onCheck={this.handleRadioTransitionChange}
                value={transition.transition}
              >
                {translate('issue.transition', transition.transition)}
              </RadioButton>
              <LightLabel>
                ({translateWithParameters('issue_bulk_change.x_issues', transition.count)})
              </LightLabel>
            </div>
          ))}
        </fieldset>
      </div>
    );
  };

  renderCommentField = () => {
    const affected = this.state.issues.filter(hasAction('comment')).length;

    if (affected === 0) {
      return null;
    }

    return (
      <FormField
        label={translate('issue.comment.formlink')}
        htmlFor="comment"
        help={
          <div className="-sw-mt-1" title={translate('issue_bulk_change.comment.help')}>
            <HelperHintIcon />
          </div>
        }
      >
        <textarea
          id="comment"
          onChange={this.handleCommentChange}
          rows={4}
          value={this.state.comment ?? ''}
        />
        <FormattingTips className="sw-text-right" />
      </FormField>
    );
  };

  renderNotificationsField = () => (
    <div>
      <Checkbox
        checked={this.state.notifications !== undefined}
        className="sw-my-2 sw-gap-1/2"
        id="send-notifications"
        onCheck={this.handleFieldCheck('notifications')}
        right
      >
        {translate('issue.send_notifications')}
      </Checkbox>
    </div>
  );

  renderForm = () => {
    const { issues, loading, paging } = this.state;

    const limitReached = paging && paging.total > MAX_PAGE_SIZE;

    return (
      <DeferredSpinner loading={loading}>
        <form id="bulk-change-form" onSubmit={this.handleSubmit}>
          {limitReached && (
            <FlagMessage className="sw-mb-4" variant="warning">
              <span>
                <FormattedMessage
                  defaultMessage={translate('issue_bulk_change.max_issues_reached')}
                  id="issue_bulk_change.max_issues_reached"
                  values={{ max: <strong>{MAX_PAGE_SIZE}</strong> }}
                />
              </span>
            </FlagMessage>
          )}

          {this.renderAssigneeField()}
          {this.renderTypeField()}
          {this.renderSeverityField()}
          {this.renderTagsField(InputField.addTags, 'issue.add_tags', true)}
          {this.renderTagsField(InputField.removeTags, 'issue.remove_tags', false)}
          {this.renderTransitionsField()}
          {this.renderCommentField()}
          {issues.length > 0 && this.renderNotificationsField()}
          {issues.length === 0 && (
            <FlagMessage variant="warning">{translate('issue_bulk_change.no_match')}</FlagMessage>
          )}
        </form>
      </DeferredSpinner>
    );
  };

  render() {
    const { issues, loading, submitting } = this.state;

    const canSubmit = this.canSubmit();

    return (
      <Modal
        onClose={this.props.onClose}
        headerTitle={
          loading
            ? translate('bulk_change')
            : translateWithParameters('issue_bulk_change.form.title', issues.length)
        }
        isScrollable
        loading={submitting}
        body={this.renderForm()}
        primaryButton={
          <ButtonPrimary
            id="bulk-change-submit"
            form="bulk-change-form"
            type="submit"
            disabled={!canSubmit || submitting || issues.length === 0}
          >
            {translate('apply')}
          </ButtonPrimary>
        }
        secondaryButtonLabel={translate('cancel')}
      />
    );
  }
}

function hasAction(action: string) {
  return (issue: Issue) => issue.actions && issue.actions.includes(action);
}

export default withOrganizationContext(BulkChangeModal);
