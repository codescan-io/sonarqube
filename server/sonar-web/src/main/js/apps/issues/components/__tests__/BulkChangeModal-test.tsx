/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { Props as ReactSelectProps } from 'react-select';
import { SelectComponentsProps } from 'react-select/src/Select';
import { searchIssueTags } from '../../../../api/issues';
import { SubmitButton } from '../../../../components/controls/buttons';
import Select, { CreatableSelect, SearchSelect } from '../../../../components/controls/Select';
import { mockIssue } from '../../../../helpers/testMocks';
import { change, waitAndUpdate } from '../../../../helpers/testUtils';
import { Issue } from '../../../../types/types';
import BulkChangeModal, { MAX_PAGE_SIZE } from '../BulkChangeModal';

jest.mock('../../../../api/issues', () => ({
  searchIssueTags: jest.fn().mockResolvedValue([undefined, []]),
}));

it('should display error message when no issues available', async () => {
  const wrapper = getWrapper([]);
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

it('should display form when issues are present', async () => {
  const wrapper = getWrapper([mockIssue()]);
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

it('should display warning when too many issues are passed', async () => {
  const issues: Issue[] = [];
  for (let i = MAX_PAGE_SIZE + 1; i > 0; i--) {
    issues.push(mockIssue());
  }

  const wrapper = getWrapper(issues);
  await waitAndUpdate(wrapper);
  expect(wrapper.find('h2')).toMatchSnapshot();
  expect(wrapper.find('Alert')).toMatchSnapshot();
});

it('should properly handle the search for tags', async () => {
  const wrapper = getWrapper([]);
  await new Promise((resolve) => {
    wrapper.instance().handleTagsSearch('query', resolve);
  });
  expect(searchIssueTags).toHaveBeenCalled();
});

it.each([
  ['type', 'set_type'],
  ['severity', 'set_severity'],
])('should render select for %s', async (_field, action) => {
  const wrapper = getWrapper([mockIssue(false, { actions: [action] })]);
  await waitAndUpdate(wrapper);

  const { Option, SingleValue } = wrapper.find<SelectComponentsProps>(Select).props().components;

  expect(Option({ data: { label: 'label', value: 'value' } })).toMatchSnapshot('Option');
  expect(SingleValue({ data: { label: 'label', value: 'value' } })).toMatchSnapshot('SingleValue');
});

it('should render tags correctly', async () => {
  const wrapper = getWrapper([mockIssue(false, { actions: ['set_tags'] })]);
  await waitAndUpdate(wrapper);

  expect(wrapper.find(CreatableSelect).exists()).toBe(true);
  expect(wrapper.find(SearchSelect).exists()).toBe(true);
});

it('should disable the submit button unless some change is configured', async () => {
  const wrapper = getWrapper([mockIssue(false, { actions: ['set_severity', 'comment'] })]);
  await waitAndUpdate(wrapper);

  return new Promise<void>((resolve) => {
    expect(wrapper.find(SubmitButton).props().disabled).toBe(true);

    // Setting a comment is not sufficient; some other change must occur.
    change(wrapper.find('#comment'), 'Some comment');
    expect(wrapper.find(SubmitButton).props().disabled).toBe(true);

    wrapper.find<ReactSelectProps>(Select).at(0).simulate('change', { value: 'foo' });

    expect(wrapper.find(SubmitButton).props().disabled).toBe(false);
    resolve();
  });
});

const getWrapper = (issues: Issue[]) => {
  return shallow<BulkChangeModal>(
    <BulkChangeModal
      component={undefined}
      currentUser={{ isLoggedIn: true, dismissedNotices: {} }}
      fetchIssues={() =>
        Promise.resolve({
          issues,
          paging: {
            pageIndex: issues.length,
            pageSize: issues.length,
            total: issues.length,
          },
        })
      }
      onClose={() => {}}
      onDone={() => {}}
    />
  );
};
