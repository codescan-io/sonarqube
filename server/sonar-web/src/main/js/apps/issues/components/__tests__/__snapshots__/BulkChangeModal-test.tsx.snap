// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display error message when no issues available 1`] = `
<Modal
  onRequestClose={[Function]}
  size="small"
>
  <form
    id="bulk-change-form"
    onSubmit={[Function]}
  >
    <div
      className="modal-head"
    >
      <h2>
        issue_bulk_change.form.title.0
      </h2>
    </div>
    <div
      className="modal-body modal-container"
    >
      <Alert
        variant="warning"
      >
        issue_bulk_change.no_match
      </Alert>
    </div>
    <div
      className="modal-foot"
    >
      <SubmitButton
        disabled={true}
        id="bulk-change-submit"
      >
        apply
      </SubmitButton>
      <ResetButtonLink
        onClick={[Function]}
      >
        cancel
      </ResetButtonLink>
    </div>
  </form>
</Modal>
`;

exports[`should display form when issues are present 1`] = `
<Modal
  onRequestClose={[Function]}
  size="small"
>
  <form
    id="bulk-change-form"
    onSubmit={[Function]}
  >
    <div
      className="modal-head"
    >
      <h2>
        issue_bulk_change.form.title.1
      </h2>
    </div>
    <div
      className="modal-body modal-container"
    >
      <Checkbox
        checked={false}
        className="display-inline-block spacer-top"
        id="send-notifications"
        onCheck={[Function]}
        right={true}
        thirdState={false}
      >
        <strong
          className="little-spacer-right"
        >
          issue.send_notifications
        </strong>
      </Checkbox>
    </div>
    <div
      className="modal-foot"
    >
      <SubmitButton
        disabled={true}
        id="bulk-change-submit"
      >
        apply
      </SubmitButton>
      <ResetButtonLink
        onClick={[Function]}
      >
        cancel
      </ResetButtonLink>
    </div>
  </form>
</Modal>
`;

exports[`should display warning when too many issues are passed 1`] = `
<h2>
  issue_bulk_change.form.title.500
</h2>
`;

exports[`should display warning when too many issues are passed 2`] = `
<Alert
  variant="warning"
>
  <FormattedMessage
    defaultMessage="issue_bulk_change.max_issues_reached"
    id="issue_bulk_change.max_issues_reached"
    values={
      {
        "max": <strong>
          500
        </strong>,
      }
    }
  />
</Alert>
`;

exports[`should render select for severity: Option 1`] = `
<Option
  data={
    {
      "label": "label",
      "value": "value",
    }
  }
>
  <SeverityHelper
    className="display-flex-center"
    severity="value"
  />
</Option>
`;

exports[`should render select for severity: SingleValue 1`] = `
<SingleValue
  data={
    {
      "label": "label",
      "value": "value",
    }
  }
>
  <SeverityHelper
    className="display-flex-center"
    severity="value"
  />
</SingleValue>
`;

exports[`should render select for type: Option 1`] = `
<Option
  data={
    {
      "label": "label",
      "value": "value",
    }
  }
>
  <div
    className="display-flex-center"
  >
    <IssueTypeIcon
      query="value"
    />
    <span
      className="little-spacer-left"
    >
      label
    </span>
  </div>
</Option>
`;

exports[`should render select for type: SingleValue 1`] = `
<SingleValue
  data={
    {
      "label": "label",
      "value": "value",
    }
  }
>
  <div
    className="display-flex-center"
  >
    <IssueTypeIcon
      query="value"
    />
    <span
      className="little-spacer-left"
    >
      label
    </span>
  </div>
</SingleValue>
`;
